### ISSUE TITLE
[Error] Cannot freeze batch norm statistics

### ISSUE THREAD

**Issue by robin-karlsson0** [open-mmlab/mmsegmentation#362]
I'm working on adversarial domain adaption experiments from within `def train_step(self, data_batch, optimizer, **kwargs)` in `mmseg/models/segmentors/encoder_decoder.py`.

However, I think I'm having **issues with not being able to freeze batch norm statistics when applying pretrained models on new samples** (i.e.` features = self.extract_feat(input_img)`)

Could you please advice on what I need to do so that models do not change between `train_step()` passes?

### What I've tried so far:

1) Set `norm_eval=True` in model config file

> norm_cfg = dict(type='BN', requires_grad=True)
> model = dict(
>     type='EncoderDecoder',
>     pretrained='open-mmlab://resnet50_v1c',
>     backbone=dict(
>         type='ResNetV1c',
>         depth=50,
>         num_stages=4,
>         out_indices=(0, 1, 2, 3),
>         dilations=(1, 1, 2, 4),
>         strides=(1, 2, 1, 1),
>         norm_cfg=norm_cfg,
>         **norm_eval=True,  # <------ HERE**
>         style='pytorch',

2) Set `model.eval()` within train.py in `mmseg/apis/train.py`
 
>     if distributed:
>         find_unused_parameters = cfg.get('find_unused_parameters', False)
>         # Sets the `find_unused_parameters` parameter in
>         # torch.nn.parallel.DistributedDataParallel
>         model = MMDistributedDataParallel(
>             model.cuda(),
>             device_ids=[torch.cuda.current_device()],
>             broadcast_buffers=False,
>             find_unused_parameters=find_unused_parameters)
>     else:
>         model = MMDataParallel(
>             model.cuda(cfg.gpu_ids[0]), device_ids=cfg.gpu_ids)
> 
>     **model.eval()  # <------ HERE**
> 
>     optimizer = build_optimizer(model, cfg.optimizer)

3) Set `backbone` and `decoder_head` to eval in `encoder_decoder.py`

> def __init__():
> ...
> self.backbone.eval()
> self.decode_head.eval()
> ...

4) Set batchnorm layers to `eval()` in `encoder_decoder.py`

> for module in self.backbone.modules():
>     if isinstance(module, nn.BatchNorm2d):
>         module.eval()

### How to reproduce:

Fine-tune a model with learning rate = 0 while outputting checkpoint files every iteration, and evaluate them using the test function. You will see that the result changes between checkpoints even though they should not (given that `norm_eval=True` would function as expected)

Thank you in advance!

**Comment by rufuzzz0** [comment]
Confirmed that I can circumvent the problem by freezing the running mean and variance in all batch norm layers, and never leave `train_step()`. Still think that freezing batch norm statics should be possible to handle in the model config file.

> def freeze_batchnorm(module):
>     '''Makes the BN running statisitcs of a module static.
>     '''
>     for module in module.modules():
>         if isinstance(module, nn.BatchNorm2d):
>             module.momentum=0.
>             module.weight.requires_grad = False
>             module.bias.requires_grad = False
>             module.eval()
> 
> class EncoderDecoder(BaseSegmentor):
>     def __init__():
> 
>         freeze_batchnorm(self.backbone)
>         freeze_batchnorm(self.decode_head)
> 
>     def train_step():
>         Do things ...

**Comment by xvjiarui** [comment]
Hi @rufuzzz0 
Sorry for the late reply.
`norm_eval` should work for `ResNet`. You may set both `lr` and `weight_decay` to zero and check if models are different.
